### ISSUE TITLE
REACT - Linking API

### ISSUE THREAD

**Issue by nfiona** [ga-wdi-exercises/project4#449]
I tried connecting [this](https://www.behance.net/v2/projects?client_id=h2dHDemBkO3RtySj6wWwnM1rXzC76TUU&field=photography) Behance API  to my search bar (like our react-OMDB homework). I expected list of projects (based on the field/ category entered in the search bar) from Behance to be displayed when a user enter a keyword. When I put a keyword in the search bar and submit, I got the attached error, instead.
My repo link is [this](https://github.com/nfiona/projectbook/blob/51dcb93826695b9235cd0302573f66f36d7eb14a/projectbook/src/components/BeUtils.js) and my question is about lines 1 - 12.

**Comment by amaseda** [comment]
Don't see an error here. Could you please paste it in here? Thx

**Comment by nfiona** [comment]
oops, here it is:

`XMLHttpRequest cannot load https://www.behance.net/v2/projects?client_id=h2dHDemBkO3RtySj6wWwnM1rXzC76TUU&field=art. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'null' is therefore not allowed access.`

**Comment by nfiona** [comment]
I did some research and it seems like a CORS issue? But I have no idea where to start to solve it..

**Comment by amaseda** [comment]
Yes, it does seem like a CORS issue.

You **might** be able to circumvent this by using `$.ajax` instead of `$.getJSON` and seeing `dataType` to `jsonp`.

https://learn.jquery.com/ajax/working-with-jsonp/

Let me know if that works or if you have any questions on set-up.

**Comment by nfiona** [comment]
Will try this and let you know. Thank you!

**Comment by amaseda** [comment]
Sounds good!

**Comment by amaseda** [comment]
Going to close this for now but feel free to re-open if you encounter any issues here!

**Comment by nfiona** [comment]
Hi @amaseda ,

The $.ajax solution solved the Access Control issue..thank you! However, now I get this error when I try submitting an input on my search bar: 
```
BeSearchContainer.js:34 Uncaught TypeError: Cannot read property 'then' of undefined
    at BeSearchContainer.onSubmitQuery (http://localhost:3000/static/js/bundle.js:33757:52)
    at handleSubmitQuery (http://localhost:3000/static/js/bundle.js:33802:28)
    at onSubmit (http://localhost:3000/static/js/bundle.js:33879:23)
    at Object.ReactErrorUtils.invokeGuardedCallback (http://localhost:3000/static/js/bundle.js:16900:17)
    at executeDispatch (http://localhost:3000/static/js/bundle.js:16682:22)
    at Object.executeDispatchesInOrder (http://localhost:3000/static/js/bundle.js:16705:6)
    at executeDispatchesAndRelease (http://localhost:3000/static/js/bundle.js:16093:23)
    at executeDispatchesAndReleaseTopLevel (http://localhost:3000/static/js/bundle.js:16104:11)
    at Array.forEach (native)
    at forEachAccumulated (http://localhost:3000/static/js/bundle.js:17003:10)
```

Line 34, repo: [BeSearchContainer.js](https://github.com/nfiona/projectbook/blob/master/projectbook/src/components/BeSearchContainer.js)


API $.ajax function:
![screen shot 2017-04-03 at 1 54 18 pm](https://cloud.githubusercontent.com/assets/24783369/24623089/47fbf280-1875-11e7-9dd4-e1eb8a186b00.png)

**Comment by amaseda** [comment]
What is the return value of `queryBehance(this.state.query)`? Is it a promise?

**Comment by nfiona** [comment]
Hi @amaseda,

I tried console.log-ing `queryBehance(this.state.query)`, and it came up as undefined. 

If I understand correctly, since my `queryBehance(query)` function's returns a value of the ajax API call (json data), then the return value of `queryBehance(this.state.query)` should not be a promise? 
Also, the ajax API call does not contain .then() method, so this does not make it a promise, right?

In other words, my answer to your question is: I'm not quite sure. I am very confused. But, if you can point me to the right direction, or you think I should use different method, I can do more research on this 😁

**Comment by amaseda** [comment]
What happens if you use `.done` instead of `.then`?

**Comment by nfiona** [comment]
same error came up unfortunately:
```
BeSearchContainer.js:35 Uncaught TypeError: Cannot read property 'done' of undefined
    at BeSearchContainer.onSubmitQuery (http://localhost:3000/static/js/bundle.js:33795:52)
    at handleSubmitQuery (http://localhost:3000/static/js/bundle.js:33840:28)
    at onSubmit (http://localhost:3000/static/js/bundle.js:33917:23)
```

**Comment by amaseda** [comment]
Does anything change if you remove the `data:` key from your AJAX call? It's a `GET` request so no need to pass data through.

**Comment by nfiona** [comment]
Yes I tried that before and it is still showing the same error..
Thank you so much for helping this late Adrian!
I found the below documentation on  Behance API JS Wrapper, do you think this may take care of the issue?
If so, I will try it tomorrow..

```
Set your Behance API key. You can set the key by either calling the export as a function or by setting a global called behance_api_key before including be.js

require(['be'], function(be) {
 be('BehanceApiKey');
});
Make the API request. The be export has various functions attached to it in order to make the API requests for the publicly available endpoints in the Behance API. Every request function returns a Promises/A+ compliant promise. They also take a callback.

 require(['be'], function(be) {
   // Using promises
   be('BehanceApiKey')
   .project.search('cats')
   .then(function success(results) {
     console.log(results);
   }, function failure(error) {
     console.error(error);
   });
 
   // Using callbacks
   be.project.search('dogs', function success(results) {
     console.log(results);
   });
 });
```

**Comment by nfiona** [comment]
I will also try using other third party APIs, if I still haven't found a solution to use Behance API by mid-day tomorrow :)
Good night!

**Comment by amaseda** [comment]
That could be it. Give it a shot!

**Comment by nfiona** [comment]
Hi Adrian!

I will close this issue for now as I am putting my third party API feature on the side and will try to set up my MERN stack now. This might also take care of the CORS issue also the API request since I will be loading axios? 🤞 
Will reopen issue when I encounter a problem :)

Thank you again for all the help!!

**Comment by amaseda** [comment]
Good call!

Axios does not support `jsonp` so if that is the problem with the Behance API, I'm afraid you'll need to use AJAX (or some other tool that allows you to set `jsonp` as the data type.

That being said, the problem could be something else so keep all approaches (including Axios) on the table.

**Comment by nfiona** [comment]
So I posted my issue on StackOverflow last night, and one person said that my queryBehance function is suffering from an [automatic semicolon insertion](http://www.bradoncode.com/blog/2015/08/26/javascript-semi-colon-insertion/), I fixed it just now and it worked! 🎊  .. 

I'm encountering a new error now but will try to solve it, and reopen this issue if I get stuck.

[My StackOverflow post](http://stackoverflow.com/questions/43196076/react-calling-third-party-api-behance-using-ajax?noredirect=1#comment73464951_43196076)

Thanks!
